### ISSUE TITLE
optimizer.minimize does not work with maxPooling layers

### ISSUE THREAD

**Issue by generic-github-user** [tensorflow/tfjs#1189]
#### TensorFlow.js version
Latest / 0.14.2

#### Browser version
Google Chrome
Version 71.0.3578.98 (Official Build) (64-bit)

#### Describe the problem

When using `optimizer.minimize()` with `model.predict()` to train a `tf.model` with a loss function, I encounter an issue. This only occurs when I use a `maxPooling2D` layer in a convolutional neural network with code similar to the code below. It produces this error: `Cannot read property 'backend' of undefined`. I'm not sure what is causing this or how to resolve it. The error does not occur when using a convolutional layer (`tf.layers.conv2d()`) without any pooling layers. I have not had this issue with `tf.layers.averagePooling2d()`, leading me to believe that this could be a possible bug.

#### Code to reproduce the error
```javascript
loss = (pred, label) => pred.sub(label).square().mean();
optimizer = tf.train.sgd(0.001);

const input = tf.input({shape: [100, 100, 4]});
const conv = tf.layers.conv2d({
	kernelSize: 5,
	filters: 8,
	strides: 1,
	activation: 'relu',
	kernelInitializer: 'VarianceScaling'
});
const pool = tf.layers.maxPooling2d({
	poolSize: [2, 2],
	strides: [2, 2]
});
const flat = tf.layers.flatten();
const dense = tf.layers.dense({units: 10});
const output = dense.apply(flat.apply(pool.apply(conv.apply(input))));
const model = tf.model({inputs: input, outputs: output});

for (var i = 0; i < 10; i++) {
	optimizer.minimize(() =>
		loss(model.predict([tf.ones([1, 100, 100, 4])]), tf.ones([1, 10]))
	);
}
```

**Comment by caisq** [comment]
@generic-github-user 

In 0.14+, there is a change that disables backpropagation support in the `Model.predict()` method. You can use the `Model.apply()` method with the `{training: true}` flag to fix your code.

```js
        optimizer.minimize(() =>
		loss(model.predict([tf.ones([1, 100, 100, 4])]), tf.ones([1, 10]))
	);
```

to

```js
       optimizer.minimize(() =>
		loss(model.apply([tf.ones([1, 100, 100, 4])], {training: true}), tf.ones([1, 10]))
	);
```

**Comment by QuantumInformation** [comment]
Initially, apply confused me as its similar name to the apply method on functions prototype.

**Comment by ichko** [comment]
This doesn't seem to work for me. Has the API been changed?
What I am trying to do is do an `optim update` using the `optim.minimize` function and a model defined with the layers API.
The `optim step` is not erroring out, but my loss is not going down.
I could not find a documentation or an example of using `apply` with this `training: true` parameter. Neither were I able to find an example of using a model defined with layers being updated with the core API (`optimizer.minimize(...)`)

Can anyone help me out?
(maybe this is not a question for this thread, but this was the only issue I was able to find that contains the setup that I am trying to recreate)
